distributed/tests/test_tls_functional.py

```diff
@@ -0,0 +1,172 @@
+
+    x = c.submit(inc, 10)
+    y = c.submit(inc, x)
+
+    result = yield c._gather(x)
+    assert result == 11
+    result = yield c._gather([x])
+    assert result == [11]
+    result = yield c._gather({'x': x, 'y': [y]})
+    assert result == {'x': 11, 'y': [12]}
+
+
+@gen_tls_cluster(client=True)
+def test_scatter(c, s, a, b):
+    assert s.address.startswith('tls://')
+
+    d = yield c._scatter({'y': 20})
+    assert s.who_has['y']
+    assert s.nbytes['y'] > 0
+    yy = yield c._gather([d['y']])
+    assert yy == [20]
+
+
+@gen_tls_cluster(client=True, Worker=Nanny)
+def test_nanny(c, s, a, b):
+    assert s.address.startswith('tls://')
+    for n in [a, b]:
+        assert isinstance(n, Nanny)
+        assert n.address.startswith('tls://')
+        assert n.worker_address.startswith('tls://')
+    assert s.ncores == {n.worker_address: n.ncores for n in [a, b]}
+
+    x = c.submit(inc, 10)
+    result = yield x._result()
+    assert result == 11
+
+
+@gen_tls_cluster(client=True)
+def test_rebalance(c, s, a, b):
+    x, y = yield c._scatter([1, 2], workers=[a.address])
+    assert len(a.data) == 2
+    assert len(b.data) == 0
+
+    yield c._rebalance()
+
+    assert len(a.data) == 1
+    assert len(b.data) == 1
+
+
+@gen_tls_cluster(client=True, ncores=[('tls://127.0.0.1', 1)] * 2)
+def test_work_stealing(c, s, a, b):
+    [x] = yield c._scatter([1], workers=a.address)
+    futures = c.map(slowadd, range(50), [x] * 50)
+    yield gen.sleep(0.1)
+    yield _wait(futures)
+    assert len(a.data) > 10
+    assert len(b.data) > 10
+    assert len(a.data) > len(b.data)
+
+
+@gen_tls_cluster(client=True)
+def test_worker_client(c, s, a, b):
+    def func(x):
+        with worker_client() as c:
+            x = c.submit(inc, x)
+            y = c.submit(double, x)
+            result = x.result() + y.result()
+            return result
+
+    x, y = c.map(func, [10, 20])
+    xx, yy = yield c._gather([x, y])
+
+    assert xx == 10 + 1 + (10 + 1) * 2
+    assert yy == 20 + 1 + (20 + 1) * 2
+
+
+@gen_tls_cluster(client=True, ncores=[('tls://127.0.0.1', 1)] * 2)
+def test_worker_client_gather(c, s, a, b):
+    a_address = a.address
+    b_address = b.address
+    assert a_address.startswith('tls://')
+    assert b_address.startswith('tls://')
+    assert a_address != b_address
+
+    def func():
+        with worker_client() as ee:
+            x = ee.submit(inc, 1, workers=a_address)
+            y = ee.submit(inc, 2, workers=b_address)
+
+            xx, yy = ee.gather([x, y])
+        return xx, yy
+
+    future = c.submit(func)
+    result = yield future._result()
+
+    assert result == (2, 3)
+
+
+@gen_tls_cluster(client=True)
+def test_worker_client_executor(c, s, a, b):
+    def mysum():
+        with worker_client() as c:
+            with c.get_executor() as e:
+                return sum(e.map(double, range(30)))
+
+    future = c.submit(mysum)
+    result = yield future._result()
+    assert result == 30 * 29
```

Comment:
I'm curious, did implementing these tests reveal issues?  I'm curious if we should anticipate other issues in the future.  What do other developers need to know when building peripheral functionality?  Are there general suggestions like "always use `self.rpc` instead of `rpc`"?